### PR TITLE
Add a pass in Analyzer for time filter optimization with preimage

### PR DESCRIPTION
**Pull request by ZhiguoZh**
The preimage has been identified as the general solution to the
optimization of predicates with time converters (https://github.com/ClickHouse/ClickHouse/issues/15257). PR https://github.com/ClickHouse/ClickHouse/pull/50951
implemented this solution via the AST rewrite.

As a follow-up, this commit extends the optimization to the experi-
mental analyzer by replacing the sub-QueryTree of the time filter
with its preimage-transformed one. The optimization is implemented
as a new pass in the Analyzer.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The performance experiments of **SSB** on the ICX device (Intel Xeon Platinum 8380 CPU, 80 cores, 160 threads) show that this change could bring an improvement of **8.5%** to the **geomean QPS** when the experimental analyzer is enabled. The details are shown below:
image

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit c933d6a9796c597d5e0be2e642575653ce19159e with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52091/c933d6a9796c597d5e0be2e642575653ce19159e/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by ZhiguoZh** [comment]
Hi @hanfei1991, with the failed cases fixed, I think this PR ready for the code review. This PR basically follows the implementation of the AST rewrite (https://github.com/ClickHouse/ClickHouse/pull/50951), and could you kindly help me review this one as well? Thanks a lot for your continuous help!

**Review by hanfei1991 — APPROVED** [review]

**Comment by ZhiguoZh** [comment]
Thanks for your review and acceptance! @hanfei1991

**Comment by novikd** [comment]
Why is this optimization not under the setting? There is no way to disable it. If it's a performance improvement why is there no new performance test?
Changing the predicate may prevent the application of read-in-order optimization.

File: src/Analyzer/Passes/OptimizeDateOrDateTimeConverterWithPreimagePass.cpp [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+                func_id = i;
+            }
+        }
+
+        if (func_id == function->getArguments().getNodes().size()) return;
+
+        size_t literal_id = 1 - func_id;
+        const auto * literal = function->getArguments().getNodes()[literal_id]->as<ConstantNode>();
+
+        if (!literal || literal->getValue().getType() != Field::Types::UInt64) return;
```
Comment (novikd) [inline src/Analyzer/Passes/OptimizeDateOrDateTimeConverterWithPreimagePass.cpp:83]: Why is the field type used?

File: src/Analyzer/Passes/OptimizeDateOrDateTimeConverterWithPreimagePass.cpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+            const auto rhs = std::make_shared<FunctionNode>("greaterOrEquals");
+            rhs->getArguments().getNodes().push_back(std::make_shared<ColumnNode>(column_node.getColumn(), column_node.getColumnSource()));
+            rhs->getArguments().getNodes().push_back(std::make_shared<ConstantNode>(end_date_or_date_time));
+            resolveOrdinaryFunctionNode(*rhs, rhs->getFunctionName());
+
+            const auto new_date_filter = std::make_shared<FunctionNode>("or");
+            new_date_filter->getArguments().getNodes() = {lhs, rhs};
+            resolveOrdinaryFunctionNode(*new_date_filter, new_date_filter->getFunctionName());
+
+            return new_date_filter;
```
Comment (novikd) [inline src/Analyzer/Passes/OptimizeDateOrDateTimeConverterWithPreimagePass.cpp:169]: Almost a complete copy-paste of `equals` case.

**Review by novikd — COMMENTED** [review]
The whole cpp file is written not like in the style guide. Very hard to read, especially because of huge amount of the copy-paste.

File: src/Analyzer/Passes/OptimizeDateOrDateTimeConverterWithPreimagePass.cpp (hunk 1 above)
Comment (ZhiguoZh) [reply src/Analyzer/Passes/OptimizeDateOrDateTimeConverterWithPreimagePass.cpp:83]: The literal in a SQL query is parsed as one of UInt64, Int64, Float64, String, Null (as [ParserLiteral](https://github.com/ClickHouse/ClickHouse/blob/master/src/Parsers/ExpressionElementParsers.h#L323) described),
and in this case, where the literal is compared with results from toYear or toYYYYMM converters, we expect the type of the literal to be UInt64, and decide to return early elsewise.

File: src/Analyzer/Passes/OptimizeDateOrDateTimeConverterWithPreimagePass.cpp (hunk 2 above)
Comment (ZhiguoZh) [reply src/Analyzer/Passes/OptimizeDateOrDateTimeConverterWithPreimagePass.cpp:169]: Thanks for pointing out this flaw! Maybe I should implement a utility function for constructing `FunctionNode` and have it build the subtrees of `equals`, `notEquals`, etc. instead, which would make the code neater.

**Comment by ZhiguoZh** [comment]
Hi Dmitry, 

I think it's required to add this optimization under the settings if it unintendedly breaks another optimization, and I will soon send out another PR for it. Meanwhile, I'd also like to explore if this preimage optimization could co-work with the read-in-order one.

And exactly, this optimization is supported more by the performance data of SSB than a CI performance test. I'm not sure if it is still necessary to add the performance test in this case, but I will try to add one if so.

The code style is also to be refined in the upcoming PR, where the code duplication problem will be further mitigated. 

Thanks for your comments! @novikd

**Comment by Algunenano** [comment]
The fuzzer is reporting a problem: https://s3.amazonaws.com/clickhouse-test-reports/0/f69bfa452633d2fcc2052a258f7202666053c787/ast_fuzzer__debug_.html

The comments from @novikid still stands by the way. The format of the file is bad, and there is no setting to disable this optimization. I'm having a look to see if it's something easy or I'll simply disable this completely until fixed

**Comment by ZhiguoZh** [comment]
Extremely sorry for failing to make progress in this code refinement in the last months due to lots of urgent internal tasks. I will try to revise the code accordingly in the coming days to 1) fix the reported problem 2) refine the code style 3) add new setting to turn the optimization on optionally. But it's still OK if it needs to be disabled temporarily before I could completely resolve the known issues.

Thanks for your comments and picking up this thread.

**Comment by Algunenano** [comment]
All 3 items should be covered by https://github.com/ClickHouse/ClickHouse/pull/60453

**Comment by ZhiguoZh** [comment]
Thanks for the fixes from which I've learned a lot. And I also noticed that the implementation of generating `FunctionNode` is not as elegant as expected. I'm working on the code refactoring locally and plan to send it out for your review after your PR gets merged.

**Comment by ZhiguoZh** [comment]
#60875 has been created for refactoring this part of code. Could you help me review it? Thanks in advance for your help!

@novikd @Algunenano

**Pull request merged**